### DISCUSSION TITLE
ofxAssimpModelLoader Update

### DISCUSSION THREAD

**Discussion by NickHardeman** [openframeworks/openFrameworks#7722] [Ideas]
Looking to put some effort towards improving ofxAssimpModelLoader. I made a list of features / functions that seem beneficial. I also tried finding relevant issues, but may have missed some. Please feel free to suggest anything or share anything that seems tricky to work with the current ofxAssimpModelLoader add-on.

General:
- [x] change to namespace approach (ofx::assimp::Model)
- [x] break out model vs model loader (model loader configures the model class)
- [x] Remove deprecated functions
- [x] add isLoaded() function #7061
- [x] extend ofNode for node manipulation capabilities
- [x] meshes are children of parent node
- [x] draw mesh by name or index #7589
- [x] get mesh with transformations applied #7303
- [x] get meshes (local or global) #5929
- [x] load via Settings struct / class similar to ofFbo::Settings. 
- [x] added centerAndScaleToWindow() function
- [x] calculation of normalized scale on animated meshes #5866
- [x] getAnimation by name
- [x] get mesh by name
- [x] get all vertices ( transformed or global )
- [x] Pass in argument to flip normals #4936
- [x] check ai Matrix to glm::mat4 conversion #1358 
- [x] Enable / Disable specific nodes ( disables rendering and animation )
- [x] Enable / disable animation updates for specific nodes

Animations
- [x] calculation of duration is incorrect in some instances; based on ticks
- [x] animations apply to specific skeletons / nodes https://github.com/openframeworks/openFrameworks/issues/7730
- [x] Nan value is calculated if duration is 0
- [x] add name for animation
- [x] Add animation from keyframe range for another animation

Mesh Helper:
- [x] getName function #3192
- [x] get mesh ( transformed or global )
- [x] Get vertices ( transformed or global )
- [x] material should be a shared_ptr so it can be modified ( https://forum.openframeworks.cc/t/how-to-clear-or-remove-a-specific-texture-from-an-ofmaterial-object/42595/4 )
- [x] Bounding box per mesh
- [x] extend ofNode per mesh; would solve #2989

Textures
- [x] Default to GL_TEXTURE_2D, can be configured on load; Will not change if OF is using arb tex globally #6218
- [x] Fix HDR images loading in emscripten relies on #7756
- [x] Fix shininess texture not being applied ( changed to apply to roughness )
- [x] Allow HDR and EXR images loading for materials.

Bones
- [x] bone structure via ofNode for familiar manipulation functions
- [x] get bone by name
- [x] get bone by index
- [x] get number of bones
- [x] return bones as vector
- [x] ability to manipulate bones procedurally with or without animation #797

Platform Specific
- [ ] android arm64-v8a https://github.com/openframeworks/openFrameworks/issues/6613
- [ ] Incorrect rendering on iOS #1723
- [x] GLTF models crashing on load for Emscripten #7738 https://github.com/openframeworks/apothecary/pull/296

Relies on 
- [x] #7766
- [x] https://github.com/openframeworks/apothecary/pull/296

**Comment by ofTheo** [comment]
Just a note that this new approach would sit side by side with the older model loader, so it wouldn't break existing code. 

One thing to figure out is: 

```
ofx::Assimp::Model
ofx::Assimp::ModelLoader 
```

vs: 

```
ofxAssimp::Model
ofxAssimp::ModelLoader etc 
```

vs: 

```
ofx::AssimpModel
ofx::AssimpModelLoader 
```

ofx could be the namespace for addons. 
but if you were to do `using namespace ofx` then assimp might clash with the lib namespace. 

curious what others think. 

cc @dimitre @artificiel @2bbb @oxillo @danoli3

**Comment by artificiel** [comment]
`ofx::assimp::Model` gets my vote — 1 namespace for all addons, then 1 namespace per add-on provides explicit granularity and contains whatever the authors put in their addons (with no extra bracing cost since C++17 nested::namespaces declarations are allowed).

about clashes, `using namespace` directives should not be defensive or boilerplate measures, but used specifically (by choice) when they substantially enhance the readability of a piece of code. in that spirit, `using namespace ofx` is a quite a wide net to cast, and you still have to write `assimp::Animation` so the unclutter is minimal — I don't see where that specific directive would be beneficial.

**Comment by 2bbb** [comment]
I agree to `ofx::asimp::Model` or `ofxAssimp::Model`.
`ofx` will be get too large namespace and be make conflict with other addons.

currently I'm keeping in mind to make addon with `ofx::Foo::Bar` style, and create alias with like `namespace ofxFoo = ofxFoo; using ofxFooBar = ofxFoo::Bar;`. and `ofx` namespace is keep clean.

**Comment by NickHardeman** [comment]
Making progress on ofxAssimp. I have a branch that still needs some work here:
https://github.com/NickHardeman/openFrameworks/tree/addon-ofxAssimp/addons/ofxAssimp

The ofx::assimp::Model class imitates the assimp scene structure and takes a node hierarchy. Each node (including the model) extends ofNode. 
There is Node, Mesh and Bone that can be accessed with get functions via name. 
`auto heelBone = skel->getNodeAsType<ofx::assimp::Bone>( "*:L_backHeel" ); 
`The asterisk before the name denotes a recursive search through the hierarchy and casts the returned Node to a Bone. 

Or you can access a node using colons as separators.  
`auto neckBone = model.getNodeAsType<ofx::assimp::Bone>("pelvis:spine:neckStart:neck_control");
`
Each mesh has a bounding box and so does the model containing all of the meshes.


https://nickhardeman.com/temp/OFExamples/ofxAssimpExample/

Not sure why the gltf models are failing and throwing an error. But it also seems to be happening with the original add-on.

**Comment by NickHardeman** [comment]
Added ability to give weights to animations and a transition function on the model class to help manage smooth transitions between animations. 
The up and down arrows control which animation and will transition between the two. 
Procedural control of bones is also working, view a sample by pressing 't' 

https://nickhardeman.com/temp/OFExamples/AssimpExample/

![image](https://github.com/openframeworks/openFrameworks/assets/142694/3ada73e9-dad4-44f9-a977-37beee7c190e)

**Reply by dimitre** [reply]
FANTASTIC!

**Reply by ggslayer** [reply]
great work!

**Comment by NickHardeman** [comment]
Pushed up an example specifically about animation and bones here:
https://github.com/NickHardeman/openFrameworks/tree/addon-ofxAssimp/examples/3d/ofxAssimpBoneControlExample

Going to focus next on an example about meshes, retrieving them, drawing them, pulling their textures, etc.

Cleaned up the example a little bit and added some additional functionality like enable / disable functionality per node.
https://nickhardeman.com/temp/OFExamples/ofxAssimpBoneControlExample/

**Comment by NickHardeman** [comment]
An example focused on moving around meshes without bones and how to get textures. 

https://nickhardeman.com/temp/OFExamples/ofxAssimpExample2/

![ofxAssimpExample](https://github.com/openframeworks/openFrameworks/assets/142694/83a5385f-7ec3-43a4-a961-8898c665c1ac)

**Comment by NickHardeman** [comment]
ofxAssimpAdvancedExample. Setting materials on meshes, 16 models using the same resources using speed and time offsets. Setting light position with offset based on bone position and orientation.

https://nickhardeman.com/temp/OFExamples/ofxAssimpAdvancedExample/

![ofxAssimpAdvancedExample](https://github.com/openframeworks/openFrameworks/assets/142694/d472a7db-fcee-48d8-a0a2-c1d1c250689b)

**Reply by dimitre** [reply]
Nick I just love your examples!

**Reply by ofTheo** [reply]
Yes!!!! Second that 😁
The staff and fox jump is just too good!!

**Comment by NickHardeman** [comment]
This is close, but I need to add some better comments to the code. 

Assimp 5.3.1 fixes the gltf loading crash, but the pull request I made to apothecary fails on the msys2 clang64 check. Seems unrelated to assimp and maybe to svgtiny, but I can't make sense of how to remedy it. 

https://github.com/openframeworks/apothecary/pull/296
